wpengine/atlas-content-modeler
# Prevent model IDs being created or registered with reserved post type slugs

Resolves #613:

- Prevents models from being registered if their ID matches post type slugs that WordPress expects to be reserved for its own use (includes 'action', 'author', 'order', 'theme', 'type', 'types').
- Disables any existing models that use those model IDs. This is to prevent unexpected behavior and fatal errors like the [one we saw in WP.org support](https://wordpress.org/support/topic/issue-when-trying-to-create-a-page/).
- Displays a message next to the disabled models:

    <img width="1437" alt="Screenshot 2022-09-14 at 20 37 10" src="https://user-images.githubusercontent.com/647669/190399732-6bb07ba0-fa22-4f37-ab6a-6c7da95339c3.png">

- Adds a link to new docs explaining how to migrate the model IDs safely. These docs depend on a WP-CLI command that does not yet exist that I will make a new PR for.

The PR also prevents creation of bad model IDs via the import tool:

<img width="993" alt="Screenshot 2022-09-14 at 14 28 40" src="https://user-images.githubusercontent.com/647669/190400153-2dd47b00-22da-446f-a9d5-14780bd8f53b.png">

And via ACM blueprints:

<img width="688" alt="Screenshot 2022-09-14 at 14 29 05" src="https://user-images.githubusercontent.com/647669/190400251-0b912ff5-95cd-4296-82f9-60cad59d719c.png">

https://wpengine.atlassian.net/browse/MTKA-1615

## Checklist

I have:


## Testing

Added tests for all new features.

### To test manually

1. Check out the `main` branch and create a model with a singular name and model ID of `type`.
2. You will see a blank page and/or fatal error when attempting to create a new post/page at `/wp-admin/post-new.php?post_type=page`
3. Check out the `fix/prevent-reserved-model-ids` branch.
4. You should be able to create pages again.
5. The Content Models → Models page should show a warning next to the model with the bad ID.

## Documentation Changes

[Preview new docs here](https://github.com/wpengine/atlas-content-modeler/blob/main/docs/help/model-id-conflicts.md).
